### PR TITLE
Coordinate derivative

### PR DESCRIPTION
**Pull request by florianwechsung**
Ufl can now (or soon) calculate shape derivatives! It's as easy as writing `derivative(J, SpatialCoordinate(mesh))`

This PR contains tests + some tiny changes to make sure the appropriate Test-/TrialFunction is created.

This needs https://bitbucket.org/fenics-project/ufl/pull-requests/90/coordinate-derivative/

### CODE REVIEW

**Comment by wence-** [comment]
Is it possible to allow the user to write either: `derivative(J, SpatialCoordinate(mesh))` or `derivative(J, mesh.coordinates)`?

Related, if I write a form that explicitly contains `mesh.coordinates`, and compute `derivative(J, SpatialCoordinate(mesh))`, do you changes DTRT?

**Comment by florianwechsung** [comment]
1. This could be added, if we can figure out by looking at a `Coefficient` whether it is actually the coordinate field of a mesh and then repalce it with `SpatialCoordinate(mesh)`

2. I'm not sure its obvious what "the right thing to do" is in that case. Currently, if you use mesh.coordinates in a form, it is treated as if it was any other Function defined on mesh, with the coordinate field interpolated into it. To illustrate: this test passes
```
mesh = UnitSquareMesh(60, 60)
V = VectorFunctionSpace(mesh, "CG", 1)
X = SpatialCoordinate(mesh)
u = Function(V)
u.interpolate(X)

J1 = inner(u, u) * dx
computed1 = assemble(derivative(J1, X)).dat.data
J2 = inner(mesh.coordinates, mesh.coordinates) * dx
computed2 = assemble(derivative(J2, X)).dat.data
assert np.allclose(computed1, computed2, rtol=1e-14)    

J3 = inner(X, X) * dx
computed3 = assemble(derivative(J3, X)).dat.data
assert not np.allclose(computed1, computed3, rtol=1e-14)    
```

This means that if you want to calculate the shape derivative for something like \int x*y*sin(x) dx, then you need to use `x, y = SpatialCoordinate(mesh)` for that.

**Comment by wence-** [comment]
Can you fix the pep8 failures in this patch?  Type `make lint` in the source tree and fix any errors.

**Comment by dham** [comment]
I think that we do want to identify `mesh.coordinates` and `SpatialCoordinate`, and that this could be done by object identity. i.e. it has to exactly be the object `mesh.coordinate`. I think that this is legitimate because `mesh.coordinate` really is the coordinate field, while `u` above is just some random field which happens to take the same value.

**Comment by florianwechsung** [comment]
Fair point. From what I see, this can be achieved in two ways:

1. On firedrake side: in the case of a `CoordinateDerivative` (or even always), replace `mesh.coordinates` with `SpatialCoordinates(mesh)`. That would involve looping over all `Coefficients` `u` in a form, checking if they are equal to `u.ufl_domain().coordinates`, and in that case replacing them with the proper `SpatialCoordinate`.
2. On UFL side: currently, if the `CoordinateRuleSet` stumbles upon a `Coefficient` `u`, it returns zero. We could add a check to see whether `u==X.ufl_domain().coordinates`, and in that case return the proper Test-/TrialFunction. That would be firedrake specific though, as in dolfin, `Mesh.coordinates` is a function that returns a numpy array.

Or is there a third option that I'm missing?

**Comment by florianwechsung** [comment]
Excellent idea, that works. I have made a small change in ufl and updated the firedrake side accordingly. 
the new test shows the behaviour

File: firedrake/ufl_expr.py [hunk 1]
```diff
@@ -138,11 +139,29 @@ def derivative(form, u, du=None, coefficient_derivatives=None):
...
+                coords = u.ufl_domain().coordinates
+            V = coords.function_space()
+            args = form.arguments()
+            number = max(a.number() for a in args) if args else -1
+            du = Argument(V, number + 1)
+            cds = {coords: du}
+            if coefficient_derivatives is not None:
+                cds.update(coefficient_derivatives)
+            coefficient_derivatives = cds
+        elif isinstance(u, firedrake.Function):
```
Comment (wence-) [inline firedrake/ufl_expr.py]: Sorry for letting this languish so long.

I think we need to provide the coefficient_derivatives in all cases where a coordinate derivative is taken.

How about this logic?

```python
def derivative(f, u, du=None, coefficient_derivatives=None):
   ...
   is_dX = isinstance(u, SpatialCoordinate) or u is f.ufl_domain().coordinates

   def argument(V):
       n = max(a.number() for a in args) if args else -1
       return Argument(V, n + 1)

   if is_dX:
       coords = f.ufl_domain().coordinates
       u = SpatialCoordinate(f.ufl_domain())
       V = coords.function_space()
       if du is None:
           du = argument(V)
       cds = {coords: du}
       if coefficient_derivatives is not None:
           cds.update(coefficient_derivatives)
       coefficient_derivatives = cds
    elif isinstance (u, ...):
        ...
```

WDYT?

Comment (florianwechsung) [reply firedrake/ufl_expr.py]: You're absolutely right, fixed it.

File: tests/regression/test_coordinatederivative.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+    V = FunctionSpace(mesh, "CG", 1)
+    u = Function(V)
+
+    J = u * u * dx
+    with pytest.raises(ValueError):
+        computed = assemble(Constant(2.0) * derivative(J, X))
+    with pytest.raises(ValueError):
+        computed = assemble(derivative(Constant(2.0) * derivative(J, X), X))
+    with pytest.raises(ValueError):
+        computed = assemble(Constant(2.0) * derivative(derivative(J, X), X))
```
Comment (thomasgibson) [inline tests/regression/test_coordinatederivative.py]: No need to assign these to anything. Just write:
```python
with pytest.raises(ValueError):
    assemble(foo)
```

**Comment by wence-** [comment]
Can you fix the flake8 failures (run `make lint` until clean)

**Comment by wence-** [comment]
And merged, thanks!

**Pull request merged**